### DISCUSSION TITLE
Add a `craft up` flag to optionally clear cache

### DISCUSSION THREAD

**Discussion by lindseydiloreto** [craftcms/cms#9846] [Ideas] [answered]
The new `craft/up` command is amazing! It perfectly combines two commands that run on every single deployment for every single site we manage. ❤️ 

Would it be possible to add a `--clear-caches` flag? That's the last missing piece of the puzzle in our standard deployment process. I believe [other folks](https://twitter.com/nystudio107/status/1085911878298148866) gravitate towards a similar workflow.

**Comment by bencroker** [comment]
While I agree that the new command, wrapped in a mutex lock, is excellent, I have to disagree with the premise that a _full_ cache clear should be part of a "standard" deployment process. The danger is that if a minor fix or update is deployed, that has no impact on the front-end, then the entire cache is busted and suddenly the server has to do a ton of work to serve the traffic. Obviously, for some sites this is fine, whereas for others it may not be, and that's the reason I challenge the suggestion – to help avoid situations in which a site experiencing high-traffic inadvertently purges its entire cache. The alternative would be add the ability to flag which caches to clear, in which case it seems easier to just run those specific extra commands.

**Reply by lindseydiloreto** [reply]
That is why I am requesting it as an _optional_ flag. 🙂 

Many people would find value in this. It is [not uncommon](https://github.com/nystudio107/craft/blob/1a2ecd3f4f969f01dc1e83fd3d38d33385b0a737/cms/composer.json#L40-L42) to clear caches during deployment.

If the flag is a bad fit for your project... just don't use it.

**Reply by bencroker** [reply]
Good point, might make sense as an _opt-in_ flag. I can just see requests for clearing specific parts of the cache following.

**Comment by brandonkelly** [comment, marked as answer]
I’d say no, because it’s not really clear which caches that’s referring to. And it’s just as easy to add a second command (e.g. `clear-caches/all`) as it is to add append a `--clear-caches` flag.

**Reply by lindseydiloreto** [reply]
Bummer. That 3-line pattern (`migrate`/`project-config`/`clear-caches`) is a really common pattern. I was hoping to whittle it down to one magical command.

**Reply by lindseydiloreto** [reply]
Unless I'm mistaken, the equivalent command would look more like this...

```
./craft up && ./craft clear-caches/all
```

Not unbearable, but I wouldn't say "just as easy".

**Reply by brandonkelly** [reply]
“Just as easy” in that you are telling it to clear caches, either way, whether that’s with a flag or a second command. And I’d argue it’s _better_ given that it’s explicitly clear which caches would be cleared as a result.

**Reply by lindseydiloreto** [reply]
Aright, fair enough. I was hoping to get a little icing on this cake, but I'll just have to be happy with `craft/up` as it is (which is already great). 🙂

Thanks @brandonkelly.

**Reply by brandonkelly** [reply]
```php
use craft\console\controllers\UpController;
use yii\base\Controller;
use yii\base\Event;

Event::on(
    UpController::class,
    Controller::EVENT_AFTER_ACTION,
    function() {
        Craft::$app->runAction('clear-caches/all');
    }
);
```

🙃

**Reply by lindseydiloreto** [reply]
Ahh, fascinating! You're implying that I can tie the commands together in a module via the `EVENT_AFTER_ACTION` event?

That might do the trick nicely... thanks for the workaround @brandonkelly! 🍺

**Reply by lindseydiloreto** [reply]
Worked like a charm, thanks again! 👍

<img src="https://user-images.githubusercontent.com/5309692/133840076-bb614202-3e63-49cd-b25b-2951a679d5bf.png" width="600">

I recognize that an Event-based workaround is not a perfect solution for all projects (or even most projects), but it's helpful for this particular site (which sits on a crappy shared server).
